utils/registry/registry.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2018 Red Hat, Inc.
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package core manages the lifecycle of all plugins (start, graceful
+// shutdown) and defines the core lifecycle SPI. The core lifecycle SPI
+// must be implemented by each plugin.
+
+package registry
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/ligato/networkservicemesh/utils/helper"
+
+	"github.com/ligato/networkservicemesh/plugins/idempotent"
+)
+
+type RegistryImpl struct {
+	plugins []idempotent.PluginAPI
+	sync.Mutex
+}
+
+// LoadOrStore - return a plugin with Deps matching p.Deps if in registry
+// or store p in the registry and return p if not
+func (r *RegistryImpl) LoadOrStore(p idempotent.PluginAPI) idempotent.PluginAPI {
+	r.Lock()
+	defer r.Unlock()
+	for _, value := range r.plugins {
+		vdeps, _ := helper.GetDeps(value)
+		pdeps, _ := helper.GetDeps(p)
+		if reflect.TypeOf(value) == reflect.TypeOf(p) && reflect.DeepEqual(pdeps, vdeps) {
+			return value
+		}
+	}
+	r.plugins = append(r.plugins, p)
+	return p
+}
+
+// Delete - delete p from registry
+func (r *RegistryImpl) Delete(p idempotent.PluginAPI) {
+	r.Lock()
+	defer r.Unlock()
+	for i, value := range r.plugins {
+		vdeps, _ := helper.GetDeps(value)
+		pdeps, _ := helper.GetDeps(p)
+		if reflect.TypeOf(value) == reflect.TypeOf(p) && reflect.DeepEqual(pdeps, vdeps) {
+			r.plugins = append(r.plugins[:i], r.plugins[i+1:]...)
+			return
+		}
+	}
+}
+
+// Clear is intended only for testing
```

Comment:
Is there an interface this is implementing? If not, perhaps we can make a test struct that extends RegistryImpl that knows how to clear the plugins slice.

Reply:
fkautz: That is a very good idea!